polyply/data/gromos_53A6_polymers/P3HT.gromos.53A6.itp

```diff
@@ -0,0 +1,198 @@
+; ;;;;; poly(3-hexyl-thiophene) (P3HT)
```

Comment:
It's probably better to make this an .ff file. It should work by just changing the extension to .ff. The benefit is that it will be treated as vermouth input file so that makes the library a bit more streamline. Also you can include a .bib file with all the citations and than add a `[ cite ]` directive like shown here: https://github.com/marrink-lab/vermouth-martinize/blob/a98098629ace902940c5e3e9dd3cd6f1ec6c0139/vermouth/data/force_fields/elnedyn/aminoacids.ff#L23

In the cite directive there should be all citations for this force-field. Alternatively you can also have a [cite] directive under the molecule in which case the citation only applies to the molecule.

Reply:
Making it a `.ff` file are re-running the above `polyply gen_itp ...` gives this:

```
Traceback (most recent call last):
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 524, in _treat_block_interaction_atoms
    reference = atom_names[reference]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/parser_utils.py", line 239, in parse_section
    return method(self, line, lineno, **kwargs)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 323, in _interactions
    _base_parser(
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 801, in _base_parser
    treated_atoms = _treat_block_interaction_atoms(atoms, context, section)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 529, in _treat_block_interaction_atoms
    raise IOError(msg.format(len(context), context.name,
OSError: There are 25 atoms defined in the block "P3HT". Interaction in section "bonds" cannot refer to atom index 31.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/riccardo/miniconda3/envs/polyply/bin/polyply", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 277, in <module>
    main()
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 273, in main
    args.func(args)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/gen_itp.py", line 53, in gen_itp
    force_field = load_library(args.name, args.lib, args.inpath)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 110, in load_library
    read_ff_from_files(all_files, force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 75, in read_ff_from_files
    wrapper(parser, path, force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 67, in wrapper
    parser(lines, force_field=force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 1003, in read_ff
    return list(director.parse(iter(lines)))
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/parser_utils.py", line 110, in parse
    result = self.dispatch(line)(line, lineno)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/parser_utils.py", line 241, in parse_section
    raise IOError("Problems parsing line {}. I think it should be a "
OSError: Problems parsing line 67. I think it should be a '['moleculetype', 'bonds']' line, but I can't parse it as such.
```

The crucial line seems to be this one?
`OSError: There are 25 atoms defined in the block "P3HT". Interaction in section "bonds" cannot refer to atom index 31.
`
as indeed the line in the itp is the one containing the atom index 31; this line is supposed to define a bond with the next monomer, so indeed 31 exceeds 25, with 25 being the number of atoms in the monomer. I thought this was the way to do it by looking at, e.g., https://github.com/marrink-lab/polyply_1.0/blob/master/polyply/data/gromos_2016H66_polymers/PEG.gromos.2016.itp or https://github.com/marrink-lab/polyply_1.0/blob/master/polyply/data/martini2_polymers/PS.martini.2.itp.

If the file is `.itp` this doesn't happen and the command successfully generates the topology of the polymer

Reply:
Regarding the citation. I tried it out by inserting the `[ citations ]` directive and by placing a `citations.bib` file there as done in the vermouth folder you linked above (neat thing, btw!) and commenting out all the lines with atom numbers > 25 from the `.ff`. If you do that, then the command runs, however, you get a new error:

```
Traceback (most recent call last):
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 74, in read_ff_from_files
    parser = FORCE_FIELD_PARSERS[file_extension]
KeyError: 'bib'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/riccardo/miniconda3/envs/polyply/bin/polyply", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 277, in <module>
    main()
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 273, in main
    args.func(args)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/gen_itp.py", line 53, in gen_itp
    force_field = load_library(args.name, args.lib, args.inpath)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 110, in load_library
    read_ff_from_files(all_files, force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 77, in read_ff_from_files
    raise IOError(
OSError: Cannot parse file with extension bib.
```
Polyply not liking to find files with extension `.bib` in there?

Reply:
@ricalessandri the bib reader is quite recent. Did you update your vermouth/martinize version already?

Reply:
Yep. I repeated the test again today, and updated the repos as well, just to double-check. I still get both errors (the 25 atoms & bib). I'm using the master/main of vermouth and the master/main of polyply.

Reply:
well in that case we have ourselves a good old dependency mess up. I'll investigate but thanks for double checking!

Reply:
#123 should fix this issues. I forgot that I didn't added the parers for citations to the polyply parsers and also there were some bugs that I had to resolve. So we'll wait for that fix and then bring through this baby

Reply:
Ok. See my feedback on https://github.com/marrink-lab/polyply_1.0/pull/123, maybe there's room for some improvements - and not sure how much of that actually depends on vermouth. After that is merged into master, I will then commit things on this PR accordingly.